src/osemgrep/targeting/Unit_semgrepignore.ml

```diff
@@ -0,0 +1,219 @@
+                   error := true)
+             else
+               match status with
+               | Not_ignored ->
+                   printf "[FAIL] %s: not ignored\n" (Git_path.to_string path);
+                   error := true
+               | Ignored ->
+                   printf "[OK] %s: ignored\n" (Git_path.to_string path));
+      assert (not !error))
+
+let tests =
+  Testutil.pack_tests "Semgrepignore"
+    [
+      ("list one file", test_list [ file "a" ]);
+      ( "list hierarchy",
+        test_list
+          [
+            file "a";
+            file "b";
+            symlink "c" "a";
+            dir "dir" [ file "d"; symlink "e" "f"; dir "g" [] ];
+          ] );
+      ( "simple gitignore",
+        test_filter
+          [ File (".gitignore", "*.c"); file "hello.c"; file "hello.ml" ]
+          [ ("/hello.ml", true); ("/hello.c", false) ] );
+      ( "relative paths",
+        test_filter
+          [ File (".gitignore", "*.c"); file "hello.c"; file "hello.ml" ]
+          [ ("hello.ml", true); ("hello.c", false) ] );
+      ( "gitignore + semgrepignore",
+        test_filter
+          [
+            File (".gitignore", "*.c");
+            File (".semgrepignore", "!hello.*");
+            file "hello.c";
+            file "hello.ml";
+          ]
+          [ ("/hello.ml", true); ("/hello.c", true); ("/generated.c", false) ]
+      );
+      ( "semgrepignore alone",
+        test_filter
+          [ File (".semgrepignore", "hello.*"); file "hello.c"; file "bye.c" ]
+          [ ("/hello.c", false); ("/bye.c", true) ] );
+      ( "deep gitignore",
+        test_filter
+          [ dir "dir" [ File (".gitignore", "a"); file "a" ]; file "a" ]
+          [ ("/a", true); ("/dir/a", false) ] );
+      ( "deep semgrepignore + gitignore",
+        test_filter
+          [
+            File (".gitignore", "a");
+            dir "dir"
+              [ File (".semgrepignore", "b"); file "a"; file "b"; file "c" ];
+            file "a";
+            file "b";
+          ]
+          [
+            ("/a", false);
+            ("/b", true);
+            ("/dir/a", false);
+            ("/dir/b", false);
+            ("/dir/c", true);
+          ] );
+      ( "ignore directories only",
+        test_filter
+          [
+            File (".gitignore", "a/");
+            dir "dir" [ file "a" ];
+            dir "a" [ file "b" ];
+          ]
+          [
+            ("/a/", false);
+            ("/a", true);
+            ("/dir/a", true);
+            ("/dir/a/", false);
+            ("/a/b", false);
+          ] );
+      ( "absolute patterns",
+        test_filter
+          [
+            File (".gitignore", "/a\nb/c");
+            dir "a" [ file "b" ];
+            dir "b" [ file "a"; file "c"; file "d"; dir "b" [ file "c" ] ];
+          ]
+          [
+            ("/a/b", false);
+            ("/b/a", true);
+            ("/b/c", false);
+            ("/b/d", true);
+            ("/b/b/c", true);
```

Comment:
so even with b/c in .gitignore, /b/b/c is not ignored?

Reply:
yes, because the pattern `b/c` must be treated as `/b/c`.

Reply:
Maybe this weird thing about b/c being converted to /b/c but foo not being convertd to /foo needs a special comment somewhere.

Reply:
Yes, I think it's a good idea to explain tricky tests.